src/Integration/Connection/ConnectionWorkflow.cs

```diff
@@ -95,28 +96,47 @@ private void OnProjectsChanged(ConnectionInformation connection, ProjectInformat
 
         #region Workflow steps
 
+        private static void AbortWorkflow(IProgressController controller, CancellationToken token)
+        {
+            bool aborted = controller.TryAbort();
+            Debug.Assert(aborted || token.IsCancellationRequested, "Failed to abort the workflow");
+        }
+
         internal /* for testing purposes */ void ConnectionStep(IProgressController controller, CancellationToken cancellationToken, ConnectionInformation connection, IProgressStepExecutionEvents notifications)
         {
             this.owner.UserNotification?.HideNotification(NotificationIds.FailedToConnectId);
+            this.owner.UserNotification?.HideNotification(NotificationIds.BadServerPluginId);
 
             notifications.ProgressChanged(connection.ServerUri.ToString(), double.NaN);
 
             ProjectInformation[] projects = this.owner.SonarQubeService.Connect(connection, cancellationToken)?.ToArray();
-            this.OnProjectsChanged(connection, projects);
 
             if (this.owner.SonarQubeService.CurrentConnection == null)
             {
                 notifications.ProgressChanged(cancellationToken.IsCancellationRequested ? Strings.ConnectionResultCancellation : Strings.ConnectionResultFailure, double.NaN);
-
-                bool aborted = controller.TryAbort();
-                Debug.Assert(aborted || cancellationToken.IsCancellationRequested, "Failed to abort the workflow");
-
                 this.owner.UserNotification?.ShowNotificationError(Strings.ConnectionFailed, NotificationIds.FailedToConnectId, this.owner.WpfCommand);
+
+                AbortWorkflow(controller, cancellationToken);
+                return;
             }
-            else
+            
+            var csPluginVersion = this.owner.SonarQubeService.GetPluginVersion(ServerPlugin.CSharpPluginKey, cancellationToken);
+            if (string.IsNullOrWhiteSpace(csPluginVersion) || VersionHelper.Compare(csPluginVersion, ServerPlugin.CSharpPluginMinimumVersion) < 0)
             {
-                notifications.ProgressChanged(Strings.ConnectionResultSuccess, double.NaN);
+                string errorMessage = string.Format(CultureInfo.CurrentCulture, Strings.ServerDoesNotHaveCorrectVersionOfCSharpPlugin, ServerPlugin.CSharpPluginMinimumVersion);
+
+                this.owner.UserNotification?.ShowNotificationError(errorMessage, NotificationIds.BadServerPluginId, this.owner.WpfCommand);
+                notifications.ProgressChanged(errorMessage, double.NaN);
```

Comment:
Ok, make sense, but consider merging those messages (otherwise you'll get two UI thread calls from the background thread).
Another issue is the ShowNotificationError(... cmd) => the message is the same, but you pass in a command (iI'm guessing to retry?), but the message itself has not command markup, and it would not make sense for the output window...